vault/request_handling.go

```diff
@@ -509,9 +563,39 @@ func (c *Core) switchedLockHandleRequest(httpCtx context.Context, req *logical.R
 		}
 	}(ctx, httpCtx)
 
-	ctx, _, req.Path, err = c.namespaceStore.ResolveNamespaceFromRequest(ctx, httpCtx, req.Path)
+	// A namespace was manually passed to HandleRequest, as can be the case with:
+	// 1. Synthesized logical requests not originating from an HTTP request
+	// 2. Tests
+	ns, err := namespace.FromContext(httpCtx)
+	// If the above is not the case, resolve the namespace from header & request path.
 	if err != nil {
-		return nil, err
+		nsHeader := namespace.HeaderFromContext(httpCtx)
+		ns, req.Path = c.namespaceStore.ResolveNamespaceFromRequest(nsHeader, req.Path)
+		if ns == nil {
+			return nil, logical.CodedError(http.StatusNotFound, "namespace not found")
+		}
+	}
+
+	isRestrictedSysAPI := ns.ID != namespace.RootNamespaceID &&
+		strings.HasPrefix(req.Path, "sys/") &&
+		restrictedSysAPIs.HasPathSegments(req.Path[4:])
+
+	if isRestrictedSysAPI {
+		return nil, logical.CodedError(http.StatusBadRequest, "operation unavailable in namespaces")
+	}
+
+	ctx = namespace.ContextWithNamespace(ctx, ns)
+
+	if ns != namespace.RootNamespace {
```

Comment:
```suggestion
	if ns.ID != namespace.RootNamespaceID {
```

We could also have one big block like this:

```go
if ns.ID != namespace.RootNamespaceID {
    // check for restricted paths
    // check for api lock
}
```

Reply:
this will probably introduce some conflicts when #1279 is merged, but why not.

Reply:
You'll probably have to rebase anyways, since #1279 isn't guaranteed to be final in its current state.